### DISCUSSION TITLE
Installation of cuQuantum on 4 A40 Datacenter GPUs for running qiskit circuits

### DISCUSSION THREAD

**Discussion by intelligi123** [NVIDIA/cuQuantum#37] [Q&A] [answered]
Hi,  
I have four A40 GPUs available which I need to make a cluster from and install cuQuantum in order to simulate some qiskit circuits to achieve parallel computing. I am getting information in bits and pieces but not sure how to start the task. 

Can someone guide me here how to first make a cluster of GPUs ready for cuQuantum installation on high level please?

**Comment by mtjrider** [comment]
> Hi, I have four A40 GPUs available which I need to make a cluster from and install cuQuantum …

What is the physical topology of the node? Are the A40s all on the same node? It would be helpful if you ran `nvidia-smi topo -m` and posted the output here. 

Regarding the second question of installing cuQuantum, I assume you mean the cuQuantum Appliance?

**Reply by intelligi123** [reply]
I have got two nodes on different locations (say A and B) and  each has 2 GPUs and I will be able to access them remotely (say C). I still have to get their access so cannot run any command right now. Can you please suggest what should be my first step once I get four IP addresses i.e. how will I convert two GPUs in one location into a single node access or how to make a cluster out of them?

Yes I mean cuQuantum Appliance.

**Reply by mtjrider** [reply]
That's helpful. Thanks. 
Do you know the interconnect between nodes? It'd be good to have some idea of available bandwidth. 

Will you be scheduling work on these nodes via resource manager (like SLURM)?

**Reply by intelligi123** [reply]
Actually, this is what I am interested to know that how can I go about as I am totally newbie to distributed computing and now have got these GPUs to work with.

You can suggest what interconnect can be used and what resource manager mostly cover general cases.

Right now what I am sure of is I have two GPUs access each with 1Tb RAM. Can you suggest a tutorial of how I can connect them to make them a cluster and then install cuQuantum?

**Reply by mtjrider** [reply]
@intelligi123 I'm sorry I completely dropped this thread.

If one of the nodes has 2x GPUs in it, and running `nvidia-smi topo -p2p rw` returns `OK` for the appropriate GPU ordinals, then all you have to do to use both GPUs collaboratively in a single workload using the container is:

`docker run -it --rm --gpus all nvcr.io/nvidia/cuquantum-appliance:xy.zw mpirun -n 2 python my_script.py`

Where `my_script.py` is either mounted into the container or available in the container. You can mount with `-v HOST_DIR:CONT_DIR`. 

An example for what `my_script.py` might look like is [here](https://docs.nvidia.com/cuda/cuquantum/latest/appliance/qiskit.html#getting-started).

Please let us know if you have other questions or need additional guidance.

**Reply by intelligi123** [reply]
Thank you @mtjrider.

I recently got the actual topology and its a sinlge server with four A40 GPUs.
I have created four VMs with a GPU each to create a distributed environment if I have to migrate GPUs in a case.

So my first step would be installation of cuquantum appliance on each VM. 
Once successful installation, i know how to run the code on one VM, my question is how to distribute it on four VMs?
Should I be using MPI or cuquantum can itself do it by getting information about other IPs?

Thank you

**Reply by mtjrider** [reply]
Can you provide more information w.r.t. how the VMs were made on the server?

Knowing more about networking and other valid data paths between GPUs helps me steer this better.

**Reply by intelligi123** [reply]
VMs have been created using VSphere with allocation of one GPU to each. Being on one server, communication bandwidth wont be any issue as far as I think but you can suggest better;
My goal is to distribute a qiskit code incrementally (on all four VMs by adding one at a time) and notice performance up gradation.

If Its possible kindly share any link of step-by-step process of installing cuquantum appliance on multiple machines and running a sample code. It would be very helpful.

**Reply by mtjrider** [reply]
Hi @intelligi123 this is a great question. I know of no previous work/test with cuQuantum + vSphere.

Roughly speaking, this is the procedure I would follow:

1. Make sure that compiled [CUDA examples](https://github.com/NVIDIA/cuda-samples/tree/master/Samples/1_Utilities/bandwidthTest) run without issue and report sensible bandwidth measures between GPU and host.
2. Configure, install, and test MPI in your virtual machines. Confirmation might look like running a simple command: `mpirun -n 2 bash -c 'echo hello from ${OMPI_COMM_WORLD_RANK}'` (assuming open-mpi).
3. [Using NCCL](https://github.com/NVIDIA/nccl-tests/blob/master/doc/PERFORMANCE.md), you can evaluate the available bandwidth between GPUs.

These items will ensure that your computational environment is well defined before you approach using cuQuantum. Unfortunately, I have limited experience with vSphere and I won't be much help if you run into problems.

**Reply by intelligi123** [reply]
Thanks for the answer and please correct me if I am wrong, Isn't OPENMPI is builtin in cuQuantum Appliance? As on its [homepage](https://catalog.ngc.nvidia.com/orgs/nvidia/containers/cuquantum-appliance) its given that for MPI applications, openMPI is already installed.

As far as I think, I need to install container (cuquantum appliance) on all four VMs and using MPI machinefile, distribute the code on vms

**Reply by mtjrider** [reply, marked as answer]
Hi @intelligi123 

Yes, `mpirun` is available in the container.
One option is to pass VM host networking through to the container, and call `mpirun` from within the container.

To do this, you will need to configure each container's SSH so that they are all mutually reachable defining aliases and hostfiles.

For example:
```
# Assuming each VM has one GPU, create a detached container named after the VM's hostname
docker run -it --detach --name client_${client_hostname}_0 --gpus "'device=0'" \
    --network=host \
    --uts=host \
    --ipc=host \
    nvcr.io/nvidia/cuquantum-appliance:23.06

# Next, create an SSH config with aliases.
# $ cat ./config
# Host client_${client_hostname}_0
#     Hostname ${client_hostname}
#     Port ${client_hostname_port}
#     IdentityFile ~/path/to/id-file-in-container
#     StrictHostKeyChecking no
#     UserKnownHostsFile=/dev/null
# ...
# $ cat ./hostfile
# client_${client_hostname}_0
# ...
mkdir -p ./.sshc && ssh-keygen -t KEY_TYPE -C '' -N '' -f ./.sshc/id_KEY_TYPE && cat ./config > ./.sshc/config && cat ./hostfile > ./.sshc/hostfile  # note my use of `.sshc`
docker cp ./sshc client_${client_hostname}_0:/container-user/home/
...
```

From here, you need to ensure that the SSH service is actually running. This may require you reconfigure the daemon/service. These commands may be useful:

```
docker exec -i client_${client_hostname}_0 bash -c 'rm -rf CONTAINER_USER_HOME_DIR/.ssh'  # remove preexisting ssh directory
docker exec -i client_${client_hostname}_0 bash -c 'mv CONTAINER_USER_HOME_DIR/.sshc CONTAINER_USER_HOME_DIR/.ssh'
docker exec -i client_${client_hostname}_0 bash -c 'chown -R {user} CONTAINER_USER_HOME_DIR/.ssh'  # fix file permissions from copy
docker exec -i client_${client_hostname}_0 bash -c 'chmod 700 CONTAINER_USER_HOME_DIR/.ssh'  # fix file permissions from copy
docker exec -i client_${client_hostname}_0 bash -c 'chmod -R 600 CONTAINER_USER_HOME_DIR/.ssh'  # fix file permissions from copy
docker exec -i client_${client_hostname}_0 bash -c \"sed -i '/Port /c\Port HOSTNAME_PORT_GOES_HERE\' /etc/ssh/sshd_config\"  # change SSH port to match SSH config alias
docker exec -i client_${client_hostname}_0 bash -c \"sed -i '/#PermitRootLogin/c\PermitRootLogin yes\' /etc/ssh/sshd_config\"  # enable root ssh login
docker exec -i client_${client_hostname}_0 service ssh restart  # restart the service
```

Once this is done on each running docker container, you should be able to launch an MPI job from within the container by passing the command the appropriate hostfile (under `CONTAINER_USER_HOME_DIR/.ssh/hostfile`).

The aforementioned gets more complicated if the networking infrastructure connecting nodes is more complex (e.g.) requiring novel drivers, etc.. The method above has the benefit of providing finely tuned control over networking interfaces. If you don't need this, you can always look into something like [mpi-operator](https://github.com/kubeflow/mpi-operator).

Let me know if you try any of the above and run into issues. I'll add that there are many ways to accomplish the same thing.

EDIT: Removed my unfortunate use of `${hostname}` as a variable when there is a command of the same name. I replaced `hostname` with `client_hostname`.

**Comment by mtjrider** [comment]
@intelligi123 
Are the questions in this discussion addressed?
Do you have any other questions for us?

**Reply by mtjrider** [reply]
@intelligi123 
Unless I hear from you, I will mark a representative response as the answer and lock this discussion. Of course, you are free to open another discussion or issue if you have other questions which need to be addressed.
